### PR TITLE
[SYCL][PI][L0] Skip modules without kernel inside when creating target kernel on L0

### PR DESCRIPTION
**Pull request by jinge90**
On L0 backend, we iterate all models in a pi_program, for each module, we call zeKernelCreate. If the call fails due to some error other than "ze invalid kernel name", the kernel creation will fail and current program will abort.
However, this logic assume that all modules in the pi_program will have some kernel inside. If we enable SYCL device library "online" link based on piProgramLink, the assumption will be broken. At that time, the pi_program will include some modules from device library without any kernel inside(they are used to resolve undefined symbol in user's image), if we call zeKernelCreate for those module, the return value is "ze invalid argument" instead of "ze invalid kernel name". Suppose under such circumstance, the pi_program includes 2 modules, mod1 for device library without any kernel and mod2 for user's device image with target kernel, we will firstly try to zeKernelCreate for mod1 and the API returns with error "ze invalid argument", then the program will abort but in fact, we can create the target kernel with mod2.
This PR will skip all modules without any kernel inside before trying to call zeKernelCreate for them, this is done by using zeModuelGetKernelNames to query the number of kernel names in module, if kernel name num is 0, we skip it.
Thanks very much.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -3491,11 +3490,19 @@ pi_result piKernelCreate(pi_program Program, const char *KernelName,
   ze_kernel_handle_t ZeKernel;
   ze_result_t ZeResult = ZE_RESULT_ERROR_INVALID_KERNEL_NAME;
   _pi_program::ModuleIterator ModIt(Program);
+  uint32_t KernelNum = 0;
```
Comment (gmlueck) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3493]: This variable is only used inside the `while` loop, so move it's declaration there.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -3491,11 +3490,19 @@ pi_result piKernelCreate(pi_program Program, const char *KernelName,
...
   while (!ModIt.Done()) {
+    KernelNum = 0;
     ZeResult =
-        ZE_CALL_NOCHECK(zeKernelCreate, (*ModIt, &ZeKernelDesc, &ZeKernel));
-    if (ZeResult != ZE_RESULT_ERROR_INVALID_KERNEL_NAME)
+        ZE_CALL_NOCHECK(zeModuleGetKernelNames, (*ModIt, &KernelNum, nullptr));
+    if (ZeResult != ZE_RESULT_SUCCESS)
       break;
```
Comment (gmlueck) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3499]: I think you could replace all of this code:

```
zResult =
    ZE_CALL_NOCHECK(zeModuleGetKernelNames, (*ModIt, &KernelNum, nullptr));
f (ZeResult != ZE_RESULT_SUCCESS)
  break;
```

with a single line:

```
ZE_CALL(zeModuleGetKernelNames, (*ModIt, &KernelNum, nullptr));
```

**Review by gmlueck — DISMISSED** [review]
LGTM

**Comment by jinge90** [comment]
Hi, @smaslov-intel 
Could you help to review this PR?
Thanks very much.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -17,16 +17,15 @@
 #include <cstdarg>
 #include <cstdio>
 #include <cstring>
+#include <level_zero/zes_api.h>
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:20]: Why did you move these?

Comment (jinge90) [reply sycl/plugins/level_zero/pi_level_zero.cpp:20]: I ran clang-format to the file before uploading the patch, it should be moved by the format tool.
Thanks very much.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -3501,10 +3500,14 @@ pi_result piKernelCreate(pi_program Program, const char *KernelName,
   ze_result_t ZeResult = ZE_RESULT_ERROR_INVALID_KERNEL_NAME;
   _pi_program::ModuleIterator ModIt(Program);
   while (!ModIt.Done()) {
-    ZeResult =
-        ZE_CALL_NOCHECK(zeKernelCreate, (*ModIt, &ZeKernelDesc, &ZeKernel));
-    if (ZeResult != ZE_RESULT_ERROR_INVALID_KERNEL_NAME)
-      break;
+    uint32_t KernelNum = 0;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3503]: Please add a source code comment about why this check is needed.

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (jinge90) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3509]: Hi, @smaslov-intel 
Comment has already been added.
Thanks very much.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:20]: can you please put them back, perhaps using comments that shut of clang-format
it is weird to have them in the middle of STL includes

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3509]: OK.

**Review by smaslov-intel — APPROVED** [review]
Just a NIT formatting ask

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (jinge90) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Hi, @smaslov-intel 
I just reverted the unnecessary header file change, could you please help approve again?
Thanks very much.

**Review by smaslov-intel — APPROVED** [review]
